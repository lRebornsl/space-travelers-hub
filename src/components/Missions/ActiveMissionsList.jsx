import { useSelector } from 'react-redux';

const ActiveMissionsList = () => {
  const missions = useSelector((state) => state.missions.missions);
  const activeMissions = missions.filter(mission => mission.reserved);

  return (
    <div>
      <h2>My Missions</h2>
      <div className='profiles-container'>
        {activeMissions.length === 0 ? (
          <p>No active missions yet.</p>
        ) : (
          <ul>
            {activeMissions.map(mission => (
              <li key={mission.mission_id}>
                {mission.mission_name}
              </li>
            ))}
          </ul>
        )}
      </div>
    </div>
  )
}

export default ActiveMissionsList;
