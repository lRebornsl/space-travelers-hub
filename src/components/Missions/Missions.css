.member-item {
  font-size: 1.6rem;
  background-color: #6d757d;
  padding: 0.4rem;
  border-radius: 0.6rem;
  color: #fefefe;
}

.mission-button {
  background: none;
  outline: 0.1rem solid black;
  border-radius: 0.25rem;
  font-size: 1.4rem;
  padding: 0 2rem;
  height: min-content;
}

.div-item {
  align-items: center;
  justify-content: center;
}

.leave-style {
  outline: 0.1rem solid red;
  color: red;
}