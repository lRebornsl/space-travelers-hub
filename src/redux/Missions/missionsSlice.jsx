import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import axios from "axios";
import { missionsURL } from "../../utils/constant";

const initialState = {
  missions: []
};

export const getMissions = createAsyncThunk("missions/getMissions", () => {
  return axios.get(missionsURL)
    .then(response => response.data)
});

const missionsSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    join: (state, action) => {
      state.missions = state.missions.map(mission => {
        if(mission.mission_id !== action.payload) 
          return mission;
        return { ...mission, reserved: true };
      });
    }
  },
  extraReducers: {
    [getMissions.fulfilled]: (state, action) => {
      if(state.missions.length === 0) {
        state.missions = action.payload;
      }
    }
  }
});

export const { join } = missionsSlice.actions;

export default missionsSlice.reducer;