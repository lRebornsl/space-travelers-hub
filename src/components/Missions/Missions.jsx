import PropTypes from 'prop-types';
import './Missions.css';
import { useDispatch } from 'react-redux';
import { join } from '../../redux/Missions/missionsSlice';

const Missions = ({ id, name, description, index, reserved }) => {
  const dispatch = useDispatch();

  const handleJoin = (e, id) => {
    e.preventDefault();
    dispatch(join(id));
  }

  return(
    <div className={`flex flex-container ${ index % 2 === 0 ? "bg-gray" : "" }`}>
      <h3 className='flex-item'>{name}</h3>
      <p className='flex-item-bigger'>{description}</p>
      <div className='flex-item flex div-item'>
        <h3 className='member-item'>NOT A MEMBER</h3>
      </div>
      <div className='flex-item flex div-item'>
        <button className={`flex-item mission-button ${ reserved ? "leave-style" : "" }`} onClick={(e) => handleJoin(e, id)}>{ reserved ? "Leave mission" : "Join Mission" }</button>
      </div>
    </div>
  )
}

Missions.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired
}

export default Missions;