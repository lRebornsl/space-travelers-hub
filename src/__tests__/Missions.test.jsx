import React from 'react';
import { render, waitFor } from '@testing-library/react';
import Missions from '../components/Missions/Missions';
import { beforeEach, describe, it } from 'vitest';
import { Provider } from 'react-redux';
import { getMissions } from '../redux/Missions/missionsSlice';
import store from '../redux/Store';

describe('Missions', () => {

  beforeEach(() => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>
    );
  });

  it('Missions renders correctly', async () => {
    await waitFor(() => expect(Missions).toMatchSnapshot());
  });

  it('Fetches missions from API', async () => {
    store.dispatch(getMissions());
    await waitFor(() => expect(store.getState().missions.missions).toHaveLength(0));
  });
});