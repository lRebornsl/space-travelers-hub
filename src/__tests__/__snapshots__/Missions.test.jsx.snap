// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Missions > Missions renders correctly 1`] = `[Function]`;
